lib/spack/spack/test/cmd/ci.py

```diff
@@ -1650,3 +1599,179 @@ def fake_dyn_mapping_urlopener(*args, **kwargs):
+          image: donotcare
+"""
+    _, output_file, _ = ci_generate_test(spack_yaml_contents)
+
+    with open(output_file) as of:
+        pipeline_doc = syaml.load(of.read())
+
+    expected_keys = ["copy", "rebuild-index", "stages", "variables", "workflow"]
+    assert all([k in pipeline_doc for k in expected_keys])
+
+    # Make sure there are only two jobs and two stages
+    stages = pipeline_doc["stages"]
+    copy_stage = "copy"
+    rebuild_index_stage = "stage-rebuild-index"
+
+    assert len(stages) == 2
+    assert stages[0] == copy_stage
+    assert stages[1] == rebuild_index_stage
+
+    rebuild_index_job = pipeline_doc["rebuild-index"]
+    assert rebuild_index_job["stage"] == rebuild_index_stage
+
+    copy_job = pipeline_doc["copy"]
+    assert copy_job["stage"] == copy_stage
+
+    # Make sure a pipeline manifest was generated
+    output_directory = os.path.dirname(output_file)
+    assert "SPACK_ARTIFACTS_ROOT" in pipeline_doc["variables"]
+    artifacts_root = pipeline_doc["variables"]["SPACK_ARTIFACTS_ROOT"]
+    pipeline_manifest_path = os.path.join(
+        output_directory, artifacts_root, "specs_to_copy", "copy_rebuilt_specs.json"
+    )
+
+    assert os.path.exists(pipeline_manifest_path)
+    assert os.path.isfile(pipeline_manifest_path)
+
+    with open(pipeline_manifest_path) as fd:
+        manifest_data = json.load(fd)
+
+    with ev.read("test") as active_env:
+        active_env.concretize()
+        for s in active_env.all_specs():
+            assert s.dag_hash() in manifest_data
+
+
+@generator("unittestgenerator")
+def generate_unittest_pipeline(
+    pipeline: PipelineDag, spack_ci: SpackCIConfig, options: PipelineOptions
+):
+    """Define a custom pipeline generator for the target 'unittestgenerator'."""
+    output_file = options.output_file
+    assert output_file is not None
+    with open(output_file, "w") as fd:
+        fd.write("unittestpipeline\n")
+        for _, (spec_label, node) in pipeline.traverse():
+            release_spec = node.spec
+            fd.write(f"  {release_spec.name}\n")
+
+
+def test_ci_generate_alternate_target(
+    ci_generate_test,
+    tmp_path,
+    mutable_mock_env_path,
+    install_mockery,
+    mock_packages,
+    ci_base_environment,
+):
+    """Ensure the above pipeline generator was correctly registered and
+    is used to generate a pipeline for the stack/config defined here."""
+    bin_mirror_url = tmp_path / "ci-bin-mirror"
+
+    spack_yaml_contents = f"""
+spack:
+  specs:
+    - archive-files
+    - externaltest
+  mirrors:
+    buildcache-destination: {bin_mirror_url}
+  ci:
+    target: unittestgenerator
+    pipeline-gen:
+    - submapping:
+      - match:
+          - archive-files
+        build-job:
+          tags:
+            - donotcare
+          image: donotcare
+"""
+    _, output_file, _ = ci_generate_test(spack_yaml_contents, "--no-prune-externals")
+
+    with open(output_file) as of:
+        pipeline_doc = syaml.load(of.read())
```

Comment:
It's definitely not valid yaml, or even yaml at all.  Why does that matter?

Reply:
Oh maybe you're saying because I used syaml to load it, instead of just reading?  In which, yeah, maybe it's confusing to load it with a yaml loader, though it doesn't seem to be a problem.

Reply:
Fixed